File: client-pc/daemon-linux/internal/daemon/daemon.go
package daemon

import (
	"context"
	"fmt"
	"time"

	"packetpilot-daemon/internal/config"
	"packetpilot-daemon/internal/logger"
	"packetpilot-daemon/internal/monitor"
	"packetpilot-daemon/internal/reporter"
)

type Daemon struct {
	config   *config.Config
	logger   *logger.Logger
	monitor  *monitor.TrafficMonitor
	reporter *reporter.Reporter
}

func New(cfg *config.Config, log *logger.Logger) (*Daemon, error) {
	// Parse durations
	updateInterval, err := time.ParseDuration(cfg.Monitor.UpdateInterval)
	if err != nil {
		return nil, fmt.Errorf("invalid update interval: %w", err)
	}

	reportInterval, err := time.ParseDuration(cfg.Reporter.ReportInterval)
	if err != nil {
		return nil, fmt.Errorf("invalid report interval: %w", err)
	}

	retryDelay, err := time.ParseDuration(cfg.Reporter.RetryDelay)
	if err != nil {
		return nil, fmt.Errorf("invalid retry delay: %w", err)
	}

	// Create monitor
	monitorConfig := &monitor.Config{
		Interface:      cfg.Monitor.Interface,
		UpdateInterval: updateInterval,
		BufferSize:     cfg.Monitor.BufferSize,
		UsageFile:      cfg.Monitor.UsageFile,
	}

	trafficMonitor := monitor.NewTrafficMonitor(monitorConfig, log)

	// Create reporter
	reporterConfig := &reporter.Config{
		ServerHost:     cfg.Server.Host,
		ServerPort:     cfg.Server.Port,
		UseTLS:         cfg.Server.UseTLS,
		APIKey:         cfg.Server.APIKey,
		DeviceID:       cfg.Server.DeviceID,
		ReportInterval: reportInterval,
		BatchSize:      cfg.Reporter.BatchSize,
		RetryAttempts:  cfg.Reporter.RetryAttempts,
		RetryDelay:     retryDelay,
	}

	trafficReporter := reporter.NewReporter(reporterConfig, log, trafficMonitor)

	return &Daemon{
		config:   cfg,
		logger:   log,
		monitor:  trafficMonitor,
		reporter: trafficReporter,
	}, nil
}

func (d *Daemon) Start(ctx context.Context) error {
	d.logger.Info("Starting PacketPilot daemon components")

	// Start traffic monitor
	go func() {
		if err := d.monitor.Start(ctx); err != nil {
			d.logger.Error("Traffic monitor failed", "error", err)
		}
	}()

	// Start reporter
	go func() {
		if err := d.reporter.Start(ctx); err != nil {
			d.logger.Error("Traffic reporter failed", "error", err)
		}
	}()

	// Wait for context cancellation
	<-ctx.Done()
	
	d.logger.Info("Shutting down daemon components")
	return nil
}

func (d *Daemon) GetDailyUsage() *monitor.DailyUsage {
	return d.monitor.GetDailyUsage()
}

func (d *Daemon) ResetStats() {
	d.monitor.ResetStats()
}

File: client-pc/daemon-win/PacketPilot.Daemon.Win/Daemon/Daemon.cs
using Microsoft.Extensions.Hosting;
using PacketPilot.Config;
using PacketPilot.Daemon.Win.Logger;
using PacketPilot.Daemon.Win.Models;
using PacketPilot.Daemon.Win.Monitor;
using PacketPilot.Daemon.Win.Reporter;
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace PacketPilot.Daemon.Win.Daemon
{
    public class Daemon : IHostedService, IDisposable
    {
        private Config.Config _config;
        private readonly Logger.Logger _logger;
        private readonly string _configPath;
        private ProcessNetworkMonitor? _processMonitor;
        private TrafficMonitor? _trafficMonitor;
        private PacketPilot.Daemon.Win.Reporter.Reporter? _reporter;
        private CancellationToken _hostCancellationToken;
        private CancellationTokenSource? _componentCancellation;
        private Task? _componentTask;
        private FileSystemWatcher? _configWatcher;
        private readonly SemaphoreSlim _reloadSemaphore = new(1, 1);
        private int _reloadScheduled;
        private bool _disposed;

        public Daemon(Config.Config config, Logger.Logger logger, ConfigPathProvider configPathProvider)
        {
            _config = config;
            _logger = logger;
            _configPath = configPathProvider.ConfigPath;
            InitializeComponents();
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _hostCancellationToken = cancellationToken;
            _logger.Info("Starting PacketPilot daemon components");
            StartConfigWatcher();
            return StartComponentsAsync();
        }

        public async Task StopAsync(CancellationToken cancellationToken)
        {
            _ = cancellationToken;
            await StopInternalAsync().ConfigureAwait(false);
        }

        public void Stop()
        {
            StopInternalAsync().GetAwaiter().GetResult();
        }

        private async Task StopInternalAsync()
        {
            StopConfigWatcher();
            await StopComponentsAsync().ConfigureAwait(false);
        }

        private void InitializeComponents()
        {
            DisposeComponents();

            _processMonitor = new ProcessNetworkMonitor(_logger, _config.Monitor);
            _trafficMonitor = new TrafficMonitor(_config.Monitor, _logger);

            var reporterConfig = new ReporterConfig
            {
                ServerHost = _config.Reporter.ServerHost,
                ServerPort = _config.Reporter.ServerPort,
                UseTls = _config.Reporter.UseTls,
                ApiKey = _config.Reporter.ApiKey,
                DeviceId = _config.Reporter.DeviceId,
                ReportInterval = _config.Reporter.ReportInterval,
                RetryAttempts = _config.Reporter.RetryAttempts,
                RetryDelay = _config.Reporter.RetryDelay,
                ReportPerProcess = _config.Reporter.ReportPerProcess
            };

            var reporterMode = _config.Reporter.ReportPerProcess ? ReporterMode.PerProcess : ReporterMode.TotalUsage;
            _reporter = new PacketPilot.Daemon.Win.Reporter.Reporter(reporterConfig, _logger, _processMonitor, _trafficMonitor, reporterMode);
        }

        private Task StartComponentsAsync()
        {
            if (_processMonitor == null || _trafficMonitor == null || _reporter == null)
            {
                throw new InvalidOperationException("Daemon components have not been initialized");
            }

            _componentCancellation = CancellationTokenSource.CreateLinkedTokenSource(_hostCancellationToken);
            var token = _componentCancellation.Token;

            var processMonitorTask = _processMonitor.StartAsync(token);
            var trafficMonitorTask = _trafficMonitor.StartAsync(token);
            var reporterTask = _reporter.StartAsync(token);

            _componentTask = Task.WhenAll(processMonitorTask, trafficMonitorTask, reporterTask);
            _ = _componentTask.ContinueWith(t =>
            {
                if (t.Exception != null)
                {
                    _logger.Error("Daemon components encountered an error", "error", t.Exception.Flatten().Message);
                }
            }, TaskContinuationOptions.OnlyOnFaulted);

            return Task.CompletedTask;
        }

        private async Task StopComponentsAsync()
        {
            if (_componentCancellation == null)
            {
                return;
            }

            _logger.Info("Stopping PacketPilot daemon components");

            try
            {
                _componentCancellation.Cancel();
            }
            catch
            {
                // Ignore cancellation errors
            }

            try
            {
                _processMonitor?.Stop();
            }
            catch (Exception ex)
            {
                _logger.Warn("Failed to stop process monitor", "error", ex.Message);
            }

            try
            {
                _trafficMonitor?.Stop();
            }
            catch (Exception ex)
            {
                _logger.Warn("Failed to stop traffic monitor", "error", ex.Message);
            }

            try
            {
                _reporter?.Stop();
            }
            catch (Exception ex)
            {
                _logger.Warn("Failed to stop reporter", "error", ex.Message);
            }

            if (_componentTask != null)
            {
                try
                {
                    await _componentTask.ConfigureAwait(false);
                }
                catch (OperationCanceledException)
                {
                    _logger.Info("Components stopped");
                }
                catch (Exception ex)
                {
                    _logger.Warn("Error while waiting for components to stop", "error", ex.Message);
                }
            }

            _componentCancellation.Dispose();
            _componentCancellation = null;
            _componentTask = null;
        }

        private void StartConfigWatcher()
        {
            var directory = Path.GetDirectoryName(_configPath);
            var fileName = Path.GetFileName(_configPath);

            if (string.IsNullOrEmpty(directory) || string.IsNullOrEmpty(fileName))
            {
                _logger.Warn("Invalid configuration path, skipping watcher setup", "path", _configPath);
                return;
            }

            _configWatcher = new FileSystemWatcher(directory, fileName)
            {
                NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.Size | NotifyFilters.FileName | NotifyFilters.CreationTime,
                EnableRaisingEvents = true
            };

            _configWatcher.Changed += HandleConfigChanged;
            _configWatcher.Created += HandleConfigChanged;
            _configWatcher.Deleted += HandleConfigChanged;
            _configWatcher.Renamed += HandleConfigChanged;
        }

        private void StopConfigWatcher()
        {
            if (_configWatcher == null)
            {
                return;
            }

            _configWatcher.EnableRaisingEvents = false;
            _configWatcher.Changed -= HandleConfigChanged;
            _configWatcher.Created -= HandleConfigChanged;
            _configWatcher.Deleted -= HandleConfigChanged;
            _configWatcher.Renamed -= HandleConfigChanged;
            _configWatcher.Dispose();
            _configWatcher = null;
        }

        private void HandleConfigChanged(object sender, FileSystemEventArgs e)
        {
            if (!string.Equals(e.FullPath, _configPath, StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            if (Interlocked.Exchange(ref _reloadScheduled, 1) == 1)
            {
                return;
            }

            _ = Task.Run(async () =>
            {
                try
                {
                    await Task.Delay(500).ConfigureAwait(false);
                    await ReloadConfigurationAsync().ConfigureAwait(false);
                }
                finally
                {
                    Interlocked.Exchange(ref _reloadScheduled, 0);
                }
            });
        }

        private async Task ReloadConfigurationAsync()
        {
            await _reloadSemaphore.WaitAsync().ConfigureAwait(false);
            try
            {
                if (_hostCancellationToken.IsCancellationRequested)
                {
                    return;
                }

                _logger.Info("Detected configuration change, reloading", "path", _configPath);
                var ensuredPath = ConfigLoader.EnsureConfigFile(_configPath);
                var latestConfig = ConfigLoader.LoadFromPath(ensuredPath);

                await StopComponentsAsync().ConfigureAwait(false);

                _config = latestConfig;
                InitializeComponents();

                if (!_hostCancellationToken.IsCancellationRequested)
                {
                    await StartComponentsAsync().ConfigureAwait(false);
                    _logger.Info("Configuration reload completed");
                }
            }
            catch (Exception ex)
            {
                _logger.Error("Failed to reload configuration", "error", ex.Message);
            }
            finally
            {
                _reloadSemaphore.Release();
            }
        }

        private void DisposeComponents()
        {
            _processMonitor?.Dispose();
            _trafficMonitor?.Dispose();
            _reporter?.Dispose();

            _processMonitor = null;
            _trafficMonitor = null;
            _reporter = null;
        }

        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            Stop();
            _reloadSemaphore.Dispose();
            DisposeComponents();
            _disposed = true;
        }
    }
}
